### DISCUSSION TITLE
Spring Boot Build Container Image on Windows Using podman

### DISCUSSION THREAD

**Discussion by kittugit** [containers/podman#14730] [Q&A]
Spring Boot 2.7.0  start to support podman to build container images.   In Windows, I try to use this feature. Have successfully installed podman. I'm able to build/run container images using Dockerfile.  When I try to build Spring Boot application, I'm not sure what could be the host address that is mentioned as follows.
Ref:  https://docs.spring.io/spring-boot/docs/current/gradle-plugin/reference/htmlsingle/#build-image.examples.docker.podman
```
tasks.named("bootBuildImage") {
	docker {
		host = "unix:///run/user/1000/podman/podman.sock"
		bindHostToBuilder = true
	}
```
}

Running " podman system connection list" gave me "ssh://user@localhost:61214/run/user/1000/podman/podman.sock".  

so when I try to build image using " .\gradlew.bat clean bootBuildImage" command, with following host url, it is throwing error 

```
tasks.named("bootBuildImage") {
	docker {
		host = "ssh://user@localhost:61214/run/user/1000/podman/podman.sock"
		bindHostToBuilder = true
	}
}
```

Error: 

Execution failed for task ':bootBuildImage'.
> Illegal char <:> at index 3: ssh://user@localhost:61214/run/user/1000/podman/podman.sock

I wonder if it is correct host address or something wrong in the gradle plugin.

**Comment by kittugit** [comment]
Using ssh port forwarding shows some progress..

 podman machine ssh podman-machine-default -L 'localhost:9999:/run/user/1000/podman/podman.sock' -N

```
tasks.named("bootBuildImage") {
	docker {
		host = "http://localhost:9999"
		bindHostToBuilder = true
	}
}
```

Getting this error:
Execution failed for task ':bootBuildImage'.
> Docker API call to 'localhost:9999/v1.24/containers/5b8c57ef2c96721189eb5e87acb1db84ccd55438f152fb5ef4f0f6f7059edf05/logs?stdout=1&stderr=1&follow=1' failed with status code 500 "Internal Server Error" and message "failed to obtain logs for Container '5b8c57ef2c96721189eb5e87acb1db84ccd55438f152fb5ef4f0f6f7059edf05': using --follow with the journald --log-driver but without the journald --events-backend (file) is not supported"

**Comment by mheon** [comment]
@n1hility Podman Machine on Windows does API socket forwarding, right? Is so, local unix socket is probably the best option.

**Reply by n1hility** [reply]
That's correct. You should not need to specify any host value here since Podman for Windows listens to the Docker Windows pipe by default. That said `podman machine start` will tell you if something else is using it (e.g. you have Docker Desktop running), and in that case it will give you a different value to use, and you can cut and paste that one into your build.

**Reply by kittugit** [reply]
@mheon @n1hility  Thank you for checking this

As suggested, I used the url from "podman maching start" in my build.gradle file and tried different option to make it work.

`API forwarding listening on: npipe:////./pipe/docker_engine`

**Attempt 1:**
Using "npipe:////./pipe/docker_engine" throws the following error:

          _Illegal char <:> at index 5: npipe:////./pipe/docker_engine_

**Attempt 2:**
Then i used  "//./pipe/docker_engine"  which shows some progress and failed with error

> Task :bootBuildImage FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':bootBuildImage'.
> Docker API call to 'localhost/v1.24/containers/create' failed with status code 500 "Internal Server Error" and message "make cli opts(): error making volume mountpoint for volume //./pipe/docker_engine: mkdir //./pipe: permission denied"_

**Attempt 3:**
 same host url ""//./pipe/docker_engine"" with rootful mode failed with following error

Execution failed for task ':bootBuildImage'.
> Docker API call to 'localhost/v1.24/containers/a2cf78e1bf12148de80ab384c02499494fbfb76a282390a6170de75ebe800ce2/logs?stdout=1&stderr=1&follow=1' failed with status code 500 "Internal Server Error" and message "failed to obtain logs for Container 'a2cf78e1bf12148de80ab384c02499494fbfb76a282390a6170de75ebe800ce2': using --follow with the journald --log-driver but without the journald --events-backend (file) is not supported"

**Comment by n1hility** [comment]
@kittugit any value it prints should go into the host property and not url, although if it is binding to docker_engine (what it attempts to do if nothing else is using it), then you don't need a host value at all. IIUC you also don't need the host property even in a custom url setting, since you can just set the DOCKER_HOST env var (something podman machine start will advise when its not using docker_engine). DOCKER_HOST should be picked up by the java client in use in gradle.
